### ISSUE TITLE
SSL errors, pipe errors, mount errors, when using with NextCloud

### ISSUE THREAD

**Issue by jakubgs** [CollaboraOnline/online#2801]
**Describe the bug**
I'm trying to use Collabora `collabora/code:6.4.10.2` Docker image with `nextcloud:21.0.3-apache` image and when I try to open a document I get a flurry of errors, and I'm not sure which ones are causing which issue and where to start.

**Expected behavior**
It works.

**Actual behavior**
It fails horribly:
```
[ docbroker_001 ] ERR  Socket #30 SSL BIO error: error:14094410:SSL routines:ssl3_read_bytes:sslv3 alert handshake failure (0: Success)| ./net/SslSocket.hpp:307
[ docbroker_001 ] ERR  Error while handling poll for socket #30 in HttpSynReqPoll: error:14094410:SSL routines:ssl3_read_bytes:sslv3 alert handshake failure| net/Socket.cpp:356
[ docbroker_001 ] ERR  WOPI::CheckFileInfo failed for URI [https://cloud.status.im/index.php/apps/wopi/wopi/files/127_oc4pzjwr04tf?access_token=f4JxQj8UYuhXsKmDbryQZYJwd9B7Rey7&access_token_ttl=1625715975000&reuse_cookies=oc_sessionPassphrase%3DFYH179rBaJT%2BDuiglgI%2FPYzSrMPH4cYqw1TLQ3%2Bl3bR7MhKGotwtxOOm9FwYR7vPANqE%2FgH5i0R2Ut3%2BzgxUJmL3xM%2FJepLhSWLZO%2B%2F3I8NyxMPEIxSJXy8Lk1bLtGtP%3Aocb1lwrdxlqi%3Ddf36fef3d95a854427900dc080d05bda%3Anc_sameSiteCookielax%3Dtrue%3Anc_sameSiteCookiestrict%3Dtrue%3Aocehkr9ykqmz%3D507e4d0adad54420a5beaf89cc452eef]: 0 . Headers: #011Body: []| wsd/Storage.cpp:723
[ docbroker_001 ] ERR  loading document exception: WOPI::CheckFileInfo failed: | wsd/DocumentBroker.cpp:1854
[ docbroker_001 ] ERR  Failed to add session to [/index.php/apps/wopi/wopi/files/127_oc4pzjwr04tf] with URI [https://cloud.status.im/index.php/apps/wopi/wopi/files/127_oc4pzjwr04tf?access_token=f4JxQj8UYuhXsKmDbryQZYJwd9B7Rey7&access_token_ttl=1625715975000&reuse_cookies=oc_sessionPassphrase%3DFYH179rBaJT%2BDuiglgI%2FPYzSrMPH4cYqw1TLQ3%2Bl3bR7MhKGotwtxOOm9FwYR7vPANqE%2FgH5i0R2Ut3%2BzgxUJmL3xM%2FJepLhSWLZO%2B%2F3I8NyxMPEIxSJXy8Lk1bLtGtP%3Aocb1lwrdxlqi%3Ddf36fef3d95a854427900dc080d05bda%3Anc_sameSiteCookielax%3Dtrue%3Anc_sameSiteCookiestrict%3Dtrue%3Aocehkr9ykqmz%3D507e4d0adad54420a5beaf89cc452eef]: WOPI::CheckFileInfo failed: | wsd/DocumentBroker.cpp:1816
[ docbroker_001 ] ERR  #25: Socket write returned -1 (EPIPE: Broken pipe)| ./net/Socket.hpp:1255
[ docbroker_001 ] ERR  Invalid or unknown session [005] to remove.| wsd/DocumentBroker.cpp:1893
[ kit_spare_002 ] ERR  mknod(/opt/lool/child-roots/yOwohX5H9tkK9WIw//tmp/dev/random) failed. Mount must not use nodev flag. (EPERM: Operation not permitted)| common/JailUtil.cpp:247
[ kit_spare_002 ] ERR  mknod(/opt/lool/child-roots/yOwohX5H9tkK9WIw//tmp/dev/urandom) failed. Mount must not use nodev flag. (EPERM: Operation not permitted)| common/JailUtil.cpp:259
[ websrv_poll ] ERR  Error while handling Client WS Request: Failed to create DocBroker with docKey [/index.php/apps/wopi/wopi/files/127_oc4pzjwr04tf].| wsd/LOOLWSD.cpp:3613
[ websrv_poll ] ERR  #29: Socket write returned -1 (EPIPE: Broken pipe)| ./net/Socket.hpp:1255
[ websrv_poll ] ERR  #29: Socket write returned -1 (EPIPE: Broken pipe)| ./net/Socket.hpp:1255
[ websrv_poll ] ERR  #29: attempted to remove: 1715 which is > size: 0 clamped to 0| ./net/Socket.hpp:1091
[ websrv_poll ] ERR  #29: Socket write returned -1 (EPIPE: Broken pipe)| ./net/Socket.hpp:1255
```
I'm seeing at least 3 distinct errors here:

* `Socket #30 SSL BIO error: error:14094410:SSL routines:ssl3_read_bytes:sslv3 alert handshake failure`
  - It appears for some reason the SSL handshake fails, but I have no idea why it even mentions `ssl3` when [it's disabled](https://github.com/CollaboraOnline/online/blob/51b87165618c1db056ebf0a41390b682787ed692/wsd/Storage.cpp#L171).
* `Socket write returned -1 (EPIPE: Broken pipe)`
  - No idea what that's referring to, might be the same SSL error, might be something else.
* ` mknod(/opt/lool/child-roots/...//tmp/dev/random) failed. Mount must not use nodev flag. (EPERM: Operation not permitted)`
  - My understanding is that this is because I'm not using `--privileged` flag as suggested [in the docs](https://sdk.collaboraonline.com/docs/installation/CODE_Docker_image.html#how-to-grab-the-code-image-from-docker-image), but they also state that this is necessary to get "faster jail creation via bind mount", so wouldn't that mean that it should also work fine without the `--privileged` flag? I have tried using `--add-cap MKNOD` but I still get these errors.

**Additional Context**

Both domains are using CloudFlare proxy with TLSv1.2 being the lowest allowed. I tried lowering that to TLSv1.0 but that didn't help with the SSL errors. My configuration sets `storage.ssl.enable` to `false`, so I'm pretty sure it should not be checking Client SSL certificates, so the failure makes no sense to me.

**Comment by jakubgs** [comment]
What confuses me is why do I see this in the logs when starting Collabora:
```
[ loolwsd ] INF  Initialized Client SSL.| wsd/Storage.cpp:182
```
When I'm using `DONT_GEN_SSL_CERT: true` as well as `storage.ssl.enable: false`. Why would it initialize it?

Code:
https://github.com/CollaboraOnline/online/blob/5a310a2232084607af9c33f7a31d61511fe47aaa/wsd/Storage.cpp#L172-L179

I have no certificates generated, and I've disabled SSL, and yet it gets configured, why?

I'm not even sure if this is the reason for my failure. But it seems related at least. The docs don't say anything about client SSL;

**Comment by jakubgs** [comment]
I tried providing [CloudFlare Client certificate](https://developers.cloudflare.com/ssl/client-certificates/create-a-client-certificate) but the error happens and is exactly the same, so the issue isn't with client SSL.

**Comment by jakubgs** [comment]
If I `curl` the domain within the container the request works just fine, no SSL issues:
```
 > d exec -it -u root nextcloud-docs bash
root@92002bf5d10e:/# curl -si https://cloud.status.im/login | grep HTTP
HTTP/2 200 
```

**Comment by tomlawesome** [comment]
> If I `curl` the domain within the container the request works just fine, no SSL issues:
> 
> ```
>  > d exec -it -u root nextcloud-docs bash
> root@92002bf5d10e:/# curl -si https://cloud.status.im/login | grep HTTP
> HTTP/2 200 
> ```

Did you resolve this problem? I'm getting the same errors, but using OwnCloud so it's related specifically to Collabora's settings/setup.

**Comment by jakubgs** [comment]
@tomlawesome nah. I'm at a complete loss. I do not understand what's happening.

It says that `Initialized Client SSL.` but I have no clue why. I'm setting `storage.ssl.enable: false` and not providing any client certs and it failed. I tried providing client certs and it failed. It makes zero sense to me. I've already wasted a LOT of time trying to debug this issue.

**Comment by tomlawesome** [comment]
> @tomlawesome nah. I'm at a complete loss. I do not understand what's happening.
> 
> It says that `Initialized Client SSL.` but I have no clue why. I'm setting `storage.ssl.enable: false` and not providing any client certs and it failed. I tried providing client certs and it failed. It makes zero sense to me. I've already wasted a LOT of time trying to debug this issue.

I went at it another way; copy loolwsd.xml out from the docket container and manually edit it, then bind mount it back in as a volume to overwrite the original. Just edit the SSL section towards the bottom of the file accordingly. 

To get rid of additional errors I had to also add:

privileged: true 

and 

cap_add:
  - MKNOD
  - SYS_CHROOT
  - FOWNER
  - CHOWN

I can at this point access the admin interface & the /hosting/discovery endpoint just fine via my collabora.domain.io but documents won’t load within my ownCloud instance.

**Comment by rugk** [comment]
…and this all looks hirribly insecure (given all the permission you need to grant)…

**Comment by tomlawesome** [comment]
> …and this all looks hirribly insecure (given all the permission you need to grant)…

Completely agree, my initial goal is just to get it operational and then to work back from there removing permissions and resolving errors one at a time.

**Comment by jakubgs** [comment]
>I went at it another way; copy loolwsd.xml out from the docket container and manually edit it, then bind mount it back in as a volume to overwrite the original. Just edit the SSL section towards the bottom of the file accordingly.

Yeah. That's exactly what I'm doing.

>…and this all looks hirribly insecure (given all the permission you need to grant)…

You can say that again. Quite worrying.

**Comment by jakubgs** [comment]
>privileged: true

I actually can't do that on my own infra because we use [user remapping](https://docs.docker.com/engine/security/userns-remap/) for additional security, but it just breaks `privileged: true`:
```
ERROR: for docs  Cannot create container for service docs:
privileged mode is incompatible with user namespaces.
You must run the container in the host namespace when running privileged mode
```

**Comment by X-dark** [comment]
I made it work with:

- extra_params=--o:ssl.enable=false --o:ssl.termination=true
- cap-add=MKNOD
- rootless podman

For some reasons, setting config.ssl.enable to false and config.ssl.termination to true in loolwsd.xml has no effect.

This is not ideal but as for me the reverse proxy (nginx) run on the same server, it is only local traffic.

**Comment by jakubgs** [comment]
I tried installing Collabora from the provided Ubuntu packages, but I'm getting the same SSL errors anyway.

At least this way I don't get any of the mounting permissions errors, but I don't know if I'm happy with running it directly on the host.

**Comment by jakubgs** [comment]
I have found a hack-fix for the SSL errors I've been seeing. I run Collabora inside Docker using Docker Compose, and this works:
```yaml
  # Fix for 'routines:ssl3_read_bytes:sslv3 alert handshake failure' 
  # Makes Collabora talk to localhost Nginx instead of CloudFlare. 
  extra_hosts: 
    cloud.status.im: 'host-gateway' 
```
This makes Collabora think it's talking directly to `cloud.example.org`, which is in my case handled by CloudFlare, but in reality it's talking directly to the Nginx reverse proxy on the local host.

**Comment by tomlawesome** [comment]
This whole issue was fixed for me by using the latest release of collabora, but also prior to that I just rolled back a couple of versions.

**Comment by jakubgs** [comment]
I'm on `6.4.10.2`, what version are you using @tomlawesome ?

**Comment by stavros-k** [comment]
> I'm on `6.4.10.2`, what version are you using @tomlawesome ?

6.4.10.6

**Comment by jakubgs** [comment]
What the hell. Actually using `6.4.10.6` did fix it for me as well. That's frustrating.

**Comment by tomlawesome** [comment]
Yeah when they released 6.4.10.6 it just started working fine, but a few days before that I rolled back one by one (docker containers) to see if it changed anything. Think it was 2-3 versions back and it was fixed for me.
